samples/switchless/README.md

```diff
@@ -0,0 +1,133 @@
+
+Prerequisite: you may want to read [Common Sample Information](../README.md#common-sample-information) before going further.
+
+## Switchless Calls
+
+In an enclave application, the host makes **ECALL**s into functions exposed by the enclaves it created. Likewise, the enclaves make **OCALL**s into functions exposed by the host that created them. In either case, the execution has to be transitioned from an untrusted environment to a trusted environment, or vice versa. Since the transition is costly due to heavy security checks, it might be more performance advantageous to make the calls **context-switchless**: the caller delegates the function call to a worker thread in the other environment, which does the real job of calling the function and post the result to the caller. Both the calling thread and the worker thread never leave their respective execution contexts during the perceived function call.
+
+The calling thread and the worker thread need to exchange information twice during the call. When the switchless call is initiated, the caller needs to pass the `job` (representing the function call) to the worker thread. And when the call finishes, the worker thread needs to pass the result back to the caller. Both exchanges need to be synchronized.
+
+## How does OE support switchless OCALLs
+
+OE only supports synchronous switchless OCALLs currently. When the caller within an enclave makes a switchless OCALL, the trusted OE runtime creates a `job` out of the function call. The `job` object includes information such as the function ID, the parameters marshaled into a buffer, and a buffer for holding the return value(s). The job is posted to a shared memory region which both the enclave and the host can access.
+
+A host worker thread checks and retrieves `job` from the shared memory region. It uses the untrusted OE runtime to process the `job` by unmarshaling the parameters, then dispatching to the callee function, and finally relaying the result back to the trusted OE runtime, which is further forwarded back to the caller.
+
+To support simultaneous switchless OCALLs made from enclaves, the host workers are multi-threaded. OE allows users to configure how many host worker threads are to be created for servicing switchless OCALLs. The following example illustrates how to do that. A word of caution is that too many host worker threads might increase competition of cores between threads and degrade the performance. Therefore, if a enclave has switchless calls enabled, OE caps the number of host worker threads for it to the number of enclave threads specified.
+
+## About the EDL
+
+First we need to define the functions we want to call between the host and the enclave. To do this we create a `switchless.edl` file:
+
+```edl
+enclave {
+    trusted {
+        public void enclave_helloworld();
+
+    };
+
+    untrusted {
+        void host_helloworld();
+        void host_helloworld_switchless() transition_using_threads;
```

Comment:
This looks like the same API is exposed both ways.  Is that what the developer should learn from this: expose every API twice?   This looks confusing as presented.